functorch/csrc/BatchingRegistrations.cpp

```diff
@@ -448,40 +448,83 @@ Tensor& zero_inplace_batching_rule(Tensor &self) {
   return self;
 }
 
-Tensor squeeze_batching_rule(const Tensor& self) {
-  if (!participatesInCurrentLevel(self)) {
-    c10::impl::ExcludeDispatchKeyGuard guard(kBatchedKey);
-    return self.squeeze();
-  }
-  auto self_physical = MultiBatchVmapTransform::logicalToPhysical(self);
-  auto physical_sizes = self_physical.tensor().sizes();
-
-  // Don't squeeze the batch dims!
-  VmapDimVector squeezed_sizes;
-  int64_t num_batch_dims = self_physical.numBatchDims();
-  squeezed_sizes.insert(
-      squeezed_sizes.end(),
-      physical_sizes.begin(),
-      physical_sizes.begin() + num_batch_dims);
-  for (auto it = physical_sizes.begin() + num_batch_dims; it != physical_sizes.end(); ++it) {
-    if (*it != 1) {
-      squeezed_sizes.push_back(*it);
+std::tuple<Tensor, optional<int64_t>> squeeze_batching_rule(const Tensor& self, optional<int64_t> bdim) {
+  if (bdim) {
+    if (self.dim() == 1) {
+      // Special case for scalar arrays to replicate PyTorch behavior.
+      return std::make_tuple<Tensor, optional<int64_t>>(
+        std::move(self.alias()),
+        std::move(bdim)
+      );
+    } else {
+      // Manually calculate the output shape by eliding all dimensions of
+      // size 1 keeping track of where the batch index started and where it
+      // ended up moving to. We also ensure we do not drop the batch index.
+
+      auto shape = self.sizes();
+      DimVector squeezed_sizes;
+      bool before_batch_idx = true;
+      int64_t new_batch_idx = 0;
+      int64_t original_idx = 0;
+
+      for (auto it = shape.begin(); it != shape.end(); ++it) {
+        // Keep only dimensions != 1 and the batch dimension (irrespective of size).
+        if (*it != 1 || original_idx == bdim) {
+          squeezed_sizes.push_back(*it);
+          if (original_idx == bdim) {
+            before_batch_idx = false;
+          }
+          // Only increment for the dimensions that will be kept in the output.
+          if (before_batch_idx) {
+            ++new_batch_idx;
+          }
+        }
+        ++original_idx;
+      }
+
+      auto result = self.view(squeezed_sizes);
+      return std::make_tuple<Tensor, optional<int64_t>>(std::move(result), c10::optional<int64_t>(new_batch_idx));
```

Comment:
nit: I think `make_tuple` can deduce the type of its template so no need to write it out: `std::make_tuple(...)`